File: ingestion/src/metadata/utils/entity_link.py
#  Copyright 2021 Collate
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  http://www.apache.org/licenses/LICENSE-2.0
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
"""
Handle Entity Link building and splitting logic.
Filter information has been taken from the
ES indexes definitions
"""
from typing import Any, List, Optional, TypeVar

from antlr4.CommonTokenStream import CommonTokenStream
from antlr4.error.ErrorStrategy import BailErrorStrategy
from antlr4.InputStream import InputStream
from antlr4.tree.Tree import ParseTreeWalker
from requests.compat import unquote_plus

from metadata.antlr.split_listener import EntityLinkSplitListener
from metadata.generated.antlr.EntityLinkLexer import EntityLinkLexer
from metadata.generated.antlr.EntityLinkParser import EntityLinkParser
from metadata.generated.schema.entity.data.table import Table
from metadata.ingestion.models.custom_pydantic import BaseModel
from metadata.utils.constants import ENTITY_REFERENCE_TYPE_MAP
from metadata.utils.dispatch import class_register

T = TypeVar("T", bound=BaseModel)


class CustomColumnName(BaseModel):
    root: str


class EntityLinkBuildingException(Exception):
    """
    Raise for inconsistencies when building the EntityLink
    """


def split(str_: str) -> List[str]:
    """
    Method to handle the splitting logic
    """

    lexer = EntityLinkLexer(InputStream(str_))
    stream = CommonTokenStream(lexer)
    parser = EntityLinkParser(stream)
    parser._errHandler = BailErrorStrategy()  # pylint: disable=protected-access
    tree = parser.entitylink()
    walker = ParseTreeWalker()
    splitter = EntityLinkSplitListener()
    walker.walk(splitter, tree)
    return splitter.split()


def get_decoded_column(entity_link: str) -> str:
    """From an URL encoded entity link get the decoded column name

    Examples:
        >>> get_decoded_column("<#E::table::rds.dev.dbt_jaffle.column_w_space::columns::first_name>")
        'first name'
        >>> get_decoded_column("<#E::table::rds.dev.dbt_jaffle.column_w_space::columns::随机的>")
        '随机的'
        >>> get_decoded_column("<#E::table::rds.dev.dbt_jaffle.table_w_space>")
        ''

    Args:
        entity_link: entity link
    """

    return CustomColumnName(
        root=unquote_plus(entity_link.split("::")[-1].replace(">", ""))
    ).root


def get_table_fqn(entity_link: str) -> str:
    """From an entity link get the table fqn

    Args:
        entity_link: entity link
    """

    return split(entity_link)[1]


def get_table_or_column_fqn(entity_link: str) -> str:
    """From an entity link get the column fqn

    Args:
        entity_link: entity link
    """
    split_entity_link = split(entity_link)
    if len(split_entity_link) == 2:
        return split_entity_link[1]
    if len(split_entity_link) == 4 and split_entity_link[2] == "columns":
        return f"{split_entity_link[1]}.{split_entity_link[3]}"

    raise ValueError(
        "Invalid entity link."
        " {split_entity_link} does not look like a table or a column entity link"
    )


get_entity_link_registry = class_register()


def get_entity_link(entity_type: Any, fqn: str, **kwargs) -> str:
    """From table fqn and column name get the entity_link

    Args:
        entity_type: Entity being built
        fqn: Entity fqn
    """

    func = get_entity_link_registry.registry.get(entity_type.__name__)
    if not func:
        return f"<#E::{ENTITY_REFERENCE_TYPE_MAP[entity_type.__name__]}::{fqn}>"

    return func(fqn, **kwargs)


@get_entity_link_registry.add(Table)
def _(fqn: str, column_name: Optional[str] = None) -> str:
    """From table fqn and column name get the entity_link"""

    if column_name:
        entity_link = f"<#E::{ENTITY_REFERENCE_TYPE_MAP[Table.__name__]}::{fqn}::columns::{column_name}>"
    else:
        entity_link = f"<#E::{ENTITY_REFERENCE_TYPE_MAP[Table.__name__]}::{fqn}>"
    return entity_link

File: openmetadata-ui/src/main/resources/ui/src/utils/EntityLink.ts
/*
 *  Copyright 2023 Collate.
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import antlr4 from 'antlr4';
import { ParseTreeWalker } from 'antlr4/src/antlr4/tree';
import EntityLinkSplitListener from '../antlr/EntityLinkSplitListener';
import { FQN_SEPARATOR_CHAR } from '../constants/char.constants';
import { FqnPart } from '../enums/entity.enum';
import EntityLinkLexer from '../generated/antlr/EntityLinkLexer';
import EntityLinkParser from '../generated/antlr/EntityLinkParser';
import { getPartialNameFromTableFQN } from './CommonUtils';
import { ENTITY_LINK_SEPARATOR } from './EntityUtils';

export default class EntityLink {
  /**
   *
   * @param string entityLink
   * @returns list of entity link parts
   */
  static split(entityLink: string) {
    if (entityLink) {
      const chars = new antlr4.InputStream(entityLink);
      const lexer = new EntityLinkLexer(chars);
      const tokens = new antlr4.CommonTokenStream(lexer);
      const parser = new EntityLinkParser(tokens);
      const tree = parser.entitylink();
      const splitter = new EntityLinkSplitListener();
      ParseTreeWalker.DEFAULT.walk(splitter, tree);

      return splitter.split();
    }

    return [];
  }

  /**
   *
   * @param string entityLink
   * @returns entity type
   */
  static getEntityType(entityLink: string) {
    return this.split(entityLink)[0];
  }

  /**
   *
   * @param string entityLink
   * @returns entity fqn
   */
  static getEntityFqn(entityLink: string) {
    return this.split(entityLink)[1];
  }

  /**
   *
   * @param string entityLink column
   * @returns entityColumn fqn
   */
  static getEntityColumnFqn(entityLink: string) {
    const parts = this.split(entityLink);

    return `${parts[1]}${parts[3] ? '.' + parts[3] : ''}`;
  }

  /**
   *
   * @param string entityLink
   * @returns entity field
   */
  static getEntityField(entityLink: string) {
    return this.split(entityLink).pop();
  }

  /**
   *
   * @param string entityLink
   * @returns column name for table entity
   */
  static getTableColumnName(entityLink: string) {
    return this.split(entityLink)[3];
  }

  /**
   *
   * @param string columnFqn
   * @param boolean withQuotes for the column name if nested column name is present
   * @returns column name for table entity
   */
  static getTableColumnNameFromColumnFqn(columnFqn: string, withQuotes = true) {
    const columnName = getPartialNameFromTableFQN(columnFqn, [
      FqnPart.NestedColumn,
    ]);

    if (!withQuotes) {
      return columnName;
    }

    return columnName.includes(FQN_SEPARATOR_CHAR)
      ? `"${columnName}"`
      : columnName;
  }

  /**
   *
   * @param string entityLink
   * @returns column field for table entity
   */
  static getTableColumnField(entityLink: string) {
    return this.split(entityLink)[4];
  }

  /**
   *
   * @param string tableFqn
   * @param string | undefined columnName
   * @returns entity link for table
   */
  static getTableEntityLink(tableFqn: string, columnName?: string) {
    if (columnName) {
      return `<#E${ENTITY_LINK_SEPARATOR}table${ENTITY_LINK_SEPARATOR}${tableFqn}${ENTITY_LINK_SEPARATOR}columns${ENTITY_LINK_SEPARATOR}${columnName}>`;
    } else {
      return `<#E${ENTITY_LINK_SEPARATOR}table${ENTITY_LINK_SEPARATOR}${tableFqn}>`;
    }
  }

  /**
   *
   * @param string entityType
   * @param string entityFqn
   * @returns entityLink
   */
  static getEntityLink(entityType: string, entityFqn: string, field?: string) {
    if (field) {
      return `<#E${ENTITY_LINK_SEPARATOR}${entityType}${ENTITY_LINK_SEPARATOR}${entityFqn}${ENTITY_LINK_SEPARATOR}${field}>`;
    }

    return `<#E${ENTITY_LINK_SEPARATOR}${entityType}${ENTITY_LINK_SEPARATOR}${entityFqn}>`;
  }
}
